File: ios/Oriveo/Oriveo/Core/Providers/GenerationParameterPresetStore.swift
import Foundation

nonisolated struct GenerationParameterPreset: Codable, Equatable, Sendable, Identifiable {
    let id: UUID
    var name: String
    let providerID: UUID
    let modelID: String
    let profileFingerprint: String
    var syncProfileKey: String? = nil
    var values: GenerationParameterOverrides
    let createdAt: Date
    var updatedAt: Date
    var revision: Int? = nil
    var mutationID: String? = nil
}

final class GenerationParameterPresetStore: @unchecked Sendable {
    static let shared = GenerationParameterPresetStore()

    private let defaults: UserDefaults
    private let key = "generation_parameter_presets.v1"
    private let lock = NSLock()

    init(defaults: UserDefaults = .standard) { self.defaults = defaults }

    /// `generation_parameter_contract.v1.json#lifecycleRules.presets`).
    func list(
        providerID: UUID,
        modelID: String,
        profileFingerprint: String,
        portableParameterIDs: Set<String> = []
    ) -> [GenerationParameterPreset] {
        records().filter { preset in
            guard preset.providerID == providerID else { return false }
            if preset.modelID == modelID { return true }
            return preset.values.values.keys.contains(where: portableParameterIDs.contains)
        }
    }

    @discardableResult
    func save(
        name: String,
        providerID: UUID,
        modelID: String,
        profileFingerprint: String,
        values: GenerationParameterOverrides,
        id: UUID? = nil
    ) -> GenerationParameterPreset {
        lock.lock()
        defer { lock.unlock() }
        var current = readLocked()
        let old = id.flatMap { candidate in current.first { $0.id == candidate } }
        let preset = GenerationParameterPreset(
            id: old?.id ?? UUID(),
            name: name.trimmingCharacters(in: .whitespacesAndNewlines),
            providerID: providerID,
            modelID: modelID,
            profileFingerprint: profileFingerprint,
            syncProfileKey: old?.syncProfileKey,
            values: values.withoutRuntimeSettings,
            createdAt: old?.createdAt ?? Date(),
            updatedAt: Date(),
            revision: (old?.revision ?? 0) + 1,
            mutationID: UUID().uuidString
        )
        current.removeAll { $0.id == preset.id }
        current.append(preset)
        GenerationParameterSyncLedger.clear(recordID: Self.recordID(preset.id), defaults: defaults)
        writeLocked(current)
        return preset
    }

    func remove(id: UUID) {
        lock.lock()
        defer { lock.unlock() }
        let current = readLocked()
        if let preset = current.first(where: { $0.id == id }) {
            GenerationParameterSyncLedger.putTombstone(
                recordID: Self.recordID(id),
                revision: (preset.revision ?? 0) + 1,
                defaults: defaults
            )
        }
        writeLocked(current.filter { $0.id != id })
    }

    func removeScopes(providerID: UUID? = nil, modelID: String? = nil) {
        guard providerID != nil || modelID != nil else { return }
        lock.lock()
        defer { lock.unlock() }
        let current = readLocked()
        let kept = current.filter { preset in
            if let providerID, preset.providerID != providerID { return true }
            if let modelID, preset.modelID != modelID { return true }
            return false
        }
        for preset in current where !kept.contains(preset) {
            GenerationParameterSyncLedger.putTombstone(
                recordID: Self.recordID(preset.id),
                revision: (preset.revision ?? 0) + 1,
                defaults: defaults
            )
        }
        writeLocked(kept)
    }

    func apply(
        _ preset: GenerationParameterPreset,
        providerID: UUID,
        modelID: String,
        profileFingerprint: String,
        semanticMapping: [String: String]? = nil
    ) -> GenerationParameterOverrides? {
        guard preset.providerID == providerID else { return nil }
        if preset.modelID == modelID { return preset.values.withoutRuntimeSettings }
        guard let semanticMapping else { return nil }
        let mapped = preset.values.values.reduce(into: [String: GenerationParameterOverride]()) { result, item in
            if let targetID = semanticMapping[item.key] { result[targetID] = item.value }
        }
        return mapped.isEmpty ? nil : GenerationParameterOverrides(values: mapped)
    }

    private func records() -> [GenerationParameterPreset] {
        lock.lock()
        defer { lock.unlock() }
        return readLocked()
    }

    private func readLocked() -> [GenerationParameterPreset] {
        guard let data = defaults.data(forKey: key) else { return [] }
        return (try? JSONDecoder().decode([GenerationParameterPreset].self, from: data)) ?? []
    }

    private func writeLocked(_ records: [GenerationParameterPreset]) {
        if let data = try? JSONEncoder().encode(records) {
            defaults.set(data, forKey: key)
        }
    }

    func syncRecords() -> [GenerationParameterPreset] { records() }

    func replaceSyncRecords(_ records: [GenerationParameterPreset]) {
        lock.lock()
        defer { lock.unlock() }
        writeLocked(records)
    }

    private static func recordID(_ id: UUID) -> String { "preset:\(id.uuidString.lowercased())" }
}

private extension GenerationParameterOverrides {
    static let runtimeParameterIDs: Set<String> = [
        "context_length", "keep_alive", "speculative_decoding", "prompt_cache", "cache_reuse",
    ]

    var withoutRuntimeSettings: GenerationParameterOverrides {
        GenerationParameterOverrides(values: values.filter { !Self.runtimeParameterIDs.contains($0.key) })
    }
}

File: android/app/src/main/java/ai/oriveo/community/core/model/GenerationParameterPresetStore.kt
package ai.oriveo.community.core.model

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import ai.oriveo.community.core.util.sameNormalizedUuid
import java.util.UUID

@Serializable
data class GenerationParameterPreset(
    val id: String,
    val name: String,
    val providerID: String,
    val modelID: String,
    val profileFingerprint: String,
    val syncProfileKey: String? = null,
    val values: GenerationParameterOverrides,
    val createdAt: Long,
    val updatedAt: Long,
    val revision: Int? = null,
    val mutationID: String? = null,
)

class GenerationParameterPresetStore(
    private val readPayload: () -> String?,
    private val writePayload: (String?) -> Unit,
    private val json: Json = Json { ignoreUnknownKeys = true },
    private val putTombstone: (String, Int) -> Unit = { _, _ -> },
    private val clearTombstone: (String) -> Unit = {},
) {

    fun list(
        providerID: String,
        modelID: String,
        profileFingerprint: String,
        portableParameterIDs: Set<String> = emptySet(),
    ): List<GenerationParameterPreset> = records().filter { preset ->

        if (!sameNormalizedUuid(preset.providerID, providerID)) return@filter false
        if (preset.modelID == modelID) return@filter true

        preset.values.values.keys.any(portableParameterIDs::contains)
    }

    fun save(
        name: String,
        providerID: String,
        modelID: String,
        profileFingerprint: String,
        values: GenerationParameterOverrides,
        id: String? = null,
    ): GenerationParameterPreset = synchronized(this) {
        val current = records().toMutableList()
        val old = id?.let { value -> current.firstOrNull { it.id == value } }
        val now = System.currentTimeMillis()
        val preset = GenerationParameterPreset(
            id = old?.id ?: UUID.randomUUID().toString(),
            name = name.trim(),
            providerID = providerID,
            modelID = modelID,
            profileFingerprint = profileFingerprint,
            syncProfileKey = old?.syncProfileKey,
            values = values.withoutRuntimeSettings(),
            createdAt = old?.createdAt ?: now,
            updatedAt = now,
            revision = (old?.revision ?: 0) + 1,
            mutationID = UUID.randomUUID().toString(),
        )
        current.removeAll { it.id == preset.id }
        current += preset
        clearTombstone(recordID(preset.id))
        writePayload(json.encodeToString(current))
        preset
    }

    fun remove(id: String) = synchronized(this) {
        val current = records()
        current.firstOrNull { it.id == id }?.let { putTombstone(recordID(id), (it.revision ?: 0) + 1) }
        writePayload(json.encodeToString(current.filterNot { it.id == id }))
    }

    fun removeScopes(providerID: String? = null, modelID: String? = null) = synchronized(this) {
        if (providerID == null && modelID == null) return@synchronized
        val current = records()
        val removed = current.filter { preset ->
            (providerID == null || sameNormalizedUuid(preset.providerID, providerID)) &&
                (modelID == null || preset.modelID == modelID)
        }
        removed.forEach { putTombstone(recordID(it.id), (it.revision ?: 0) + 1) }
        writePayload(json.encodeToString(current.filterNot { it in removed }))
    }

    fun clearForAccountBoundary() = synchronized(this) { writePayload(null) }

    fun apply(
        preset: GenerationParameterPreset,
        providerID: String,
        modelID: String,
        profileFingerprint: String,
        semanticMapping: Map<String, String>? = null,
    ): GenerationParameterOverrides? {
        if (!sameNormalizedUuid(preset.providerID, providerID)) return null
        if (preset.modelID == modelID) return preset.values.withoutRuntimeSettings()
        val mapping = semanticMapping ?: return null
        val mapped = preset.values.values.mapNotNull { (sourceID, value) -> mapping[sourceID]?.let { it to value } }.toMap()
        return mapped.takeIf { it.isNotEmpty() }?.let(::GenerationParameterOverrides)
    }

    private fun records(): List<GenerationParameterPreset> = readPayload()?.let { payload ->
        runCatching { json.decodeFromString<List<GenerationParameterPreset>>(payload) }.getOrDefault(emptyList())
    } ?: emptyList()

    internal fun syncRecords(): List<GenerationParameterPreset> = synchronized(this) { records() }

    internal fun replaceSyncRecords(records: List<GenerationParameterPreset>) = synchronized(this) {
        writePayload(json.encodeToString(records))
    }

    private fun recordID(id: String): String = "preset:${id.lowercase()}"

    private fun GenerationParameterOverrides.withoutRuntimeSettings() = GenerationParameterOverrides(
        values.filterKeys { it !in RUNTIME_PARAMETER_IDS },
    )

    companion object {
        internal const val PREFS_NAME = "generation_parameter_presets"
        internal const val KEY_PAYLOAD = "v1"
        private val RUNTIME_PARAMETER_IDS = setOf("context_length", "keep_alive", "speculative_decoding", "prompt_cache", "cache_reuse")

        fun from(context: Context): GenerationParameterPresetStore {
            val prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            return GenerationParameterPresetStore(
                readPayload = { prefs.getString(KEY_PAYLOAD, null) },
                writePayload = { payload -> prefs.edit().putString(KEY_PAYLOAD, payload).apply() },
                putTombstone = { recordID, revision -> GenerationParameterSyncLedger.from(context).put(recordID, revision) },
                clearTombstone = { recordID -> GenerationParameterSyncLedger.from(context).clear(recordID) },
            )
        }
    }
}
